en_US/advanced/lang-exproto.md

```diff
@@ -0,0 +1,144 @@
+---
+# 编写日期
+date: 2021-03-09 13:56:26
+# 作者 Github 名称
+author: hjianbo
+# 关键字
+keywords:
+# 描述
+description:
+# 分类
+category: 
+# 引用
+ref:
+---
+
+# Extension Protocol
+
+**Extension Protocol** is provided by the **emqx-exproto** plugin, which allows other programming languages (e.g. Python, Java, etc.) to process bytes directly for parsing private protocols and provides a Pub/Sub interface for message exchange with EMQ X.
+
+This feature gives EMQ X the power of scalability to handle any private protocol in a user-friendly programming language and enjoy the benefits of extremely high concurrent connections brought by the EMQ X.
+
+
+## Features
+
+- Extremely scalable. Supports all major programming languages using gRPC as the RPC communication framework
+- Fully asynchronous IO. the connection layer is implemented as fully asynchronous non-blocking I/O
+- Transparent connection layer. Full support for TCP/TLS UDP/DTLS type connection management, and provides a unified API for the upper layers
+- Connection management capabilities. For example, maximum number of connections, connection and throughput rate limits, IP blacklisting, etc.
+
+
+## Design
+
+![Extension-Protocol Arch](./assets/exproto-arch.jpg)
+
+The interfaces provided by emqx-exproto:
+
+1. **Connection Layer:** This layer mainly **maintains the life cycle of the socket, and the sending/receiving of data**. Included:
+    - Listens on a port. When a new TCP/UDP connection arrives, a connection process is started to maintain the state of the connection.
+    - Call the `OnSocketCreated` callback. Used to notify the user's server that **a new connection has been established**.
+    - Call `OnSocketClosed` callback. Used to notify the user's server that a connection **has been closed**.
+    - Call `OnReceivedBytes` callback. Used to notify the user's server that **the connection received new packets**.
+    - Provides the `Send` interface. Called by user's server to **send packets**.
+    - Provides the `Close` interface. Called by user's server. **For actively closing the connection**.
+
+2. **Protocol/Session Layer:** This layer primarily **provides the PUB/SUB interface** for message interoperability with the EMQ X. Includes:
+    - Provides the `Authenticate` interface. Called by user's server to register clients into EMQ X.
+    - Provides the `StartTimer` interface. Called by user's server to start a timer such as a heartbeat for the connected process.
+    - Provides the `Publish` interface. Called by user's server to publish messages to the EMQ X.
+    - Provides the `Subscribe` interface. Called by user's server to subscribe to a topic to receive certain downlink messages from the EMQ X.
+    - Provides the `Unsubscribe` interface. Called by user's server to unsubscribe from a topic.
+    - Calls the `OnTimerTimeout` callback. Used to handle timer timeout events
+    - Call the `OnReceivedMessages` callback. Used to receive downlink messages.
```

Comment:
Used to receive downlink messages. -> Used to receive downlink messages (After a successful subscription to a topic, this method will be called back if there are messages on the topic)